src/Routing/Duplex.purs

```diff
@@ -64,40 +72,116 @@ instance applicativeRouteDuplex :: Applicative (RouteDuplex i) where
 
+-- | Similar to `prefix`, strips (when parsing) or adds (when printing) given string segment of the path.
+-- | Similar precautions as for prefix apply here.
 suffix :: forall a b. RouteDuplex a b -> String -> RouteDuplex a b
 suffix (RouteDuplex enc dec) s = RouteDuplex (\a -> enc a <> Printer.put s) (dec <* Parser.prefix s (pure unit))
 
+-- | Strips (when parsing) or adds (when printing) given String prefix,
+-- | potentially consisting of multiple path segments. Constrast this with `prefix`,
+-- | which only deal with single segment.
+-- |
+-- |```purescript
+-- | parse (path "/api/v1" segment) "/api/v1/a" == Right "a"
+-- | parse (path "/api/v1" segment) "/api/v2/a" == Left (Expected "v1" "v2")
+-- |```
 path :: forall a b. String -> RouteDuplex a b -> RouteDuplex a b
 path = flip (foldr prefix) <<< String.split (Pattern "/")
 
+-- | Modifies given codec by requiring it to be prefixed with '/'.
+-- | You can think of it as stripping the '/' at the beginning of path,
+-- | failing if it's not there.
+-- |
+-- |```purescript
+-- | parse (root segment) "/abc" == Right "abc"
+-- | parse (root segment) "abc" == Left (Expected "" "abc")
+-- |
+-- | print (root segment) "abc" == "/abc"
+-- |```
 root :: forall a b. RouteDuplex a b -> RouteDuplex a b
 root = path ""
 
+-- | `end codec` will only suceed if codec succeeds and there are no
+-- | additional path segments remaining to be processed.
+-- |
+-- |```purescript
+-- | parse (end segment) "abc" == Right "abc"
+-- | parse (end segment) "abc/def" == Left (ExpectedEndOfPath "def")
+-- |```
 end :: forall a b. RouteDuplex a b -> RouteDuplex a b
 end (RouteDuplex enc dec) = RouteDuplex enc (dec <* Parser.end)
 
+-- | Consumes or prints a single path segment.
+-- | Note that uri encoding / decoding is done automatically.
+-- |
+-- | ```purescript
+-- | parse segment "abc"         == Right "abc"
+-- | parse segment "abc%20def"   == Right "abc def" -- automatic decoding of uri components
+-- | parse segment "abc/def"     == Right "abc"
+-- | parse segment "/abc"        == Right "" -- the empty string before the first '/'
+-- | parse (root segment) "/abc" == Right "abc"
+-- |
+-- | print segment "hello there" == "hello%20there"
+-- | print segment "" == "/"
+-- | ```
 segment :: RouteDuplex' String
 segment = RouteDuplex Printer.put Parser.take
 
+-- | `param name` consumes or prints a query parameter with given name.
+-- | Parsing will fail if the parameter is not there.
+-- |
+-- |```purescript
+-- | parse (param "search") "?search=keyword" == Right "keyword"
+-- | parse (param "search") "/"               == Left (MissingParam "search")
+-- | parse (optional (param "search")) "/"    == Right Nothing
+-- |```
 param :: String -> RouteDuplex' String
 param p = RouteDuplex (Printer.param p) (Parser.param p)
 
+-- TODO not sure what's use case for this combinator? Is it like part of path that can optionally be present? Or is it about query params?
 flag :: RouteDuplex' String -> RouteDuplex' Boolean
```

Comment:
Just one question: It's not clear to me what would be the use case of this combinator..

Reply:
This is for query parameters like "?flag", where it's presence indicates true, and it's absence indicates false. Contrast with `boolean (param "flag")` which requires it to be `?flag=true` or `?flag=false`, and absence fails to parse.